torchreid/engine/image/am_softmax.py

```diff
@@ -166,62 +192,71 @@ def forward_backward(self, data):
-                    out_logits[trg_id].append(model_logits[trg_id])
+                for trg_id in range(self.num_targets):
+                    if model_logits[trg_id] is not None:
+                        out_logits[trg_id].append(model_logits[trg_id])
 
-        if len(model_names) > 1:
-            num_mutual_losses = 0
-            mutual_loss = torch.zeros([], dtype=imgs.dtype, device=imgs.device)
-            for trg_id in range(self.num_targets):
-                if len(out_logits[trg_id]) <= 1:
-                    continue
+            if len(model_names) > 1:
+                num_mutual_losses = 0
+                mutual_loss = torch.zeros([], dtype=imgs.dtype, device=imgs.device)
+                for trg_id in range(self.num_targets):
+                    if len(out_logits[trg_id]) <= 1:
+                        continue
 
-                with torch.no_grad():
-                    trg_probs = torch.softmax(torch.stack(out_logits[trg_id]), dim=2).mean(dim=0)
-
-                for model_id, logits in enumerate(out_logits[trg_id]):
-                    log_probs = torch.log_softmax(logits, dim=1)
-                    m_loss = (trg_probs * log_probs).sum(dim=1).mean().neg()
+                    with torch.no_grad():
+                        trg_probs = torch.softmax(torch.stack(out_logits[trg_id]), dim=2).mean(dim=0)
 
-                    mutual_loss += m_loss
-                    loss_summary['mutual_{}/{}'.format(trg_id, model_names[model_id])] = m_loss.item()
-                    num_mutual_losses += 1
+                    for model_id, logits in enumerate(out_logits[trg_id]):
+                        log_probs = torch.log_softmax(logits, dim=1)
+                        m_loss = (trg_probs * log_probs).sum(dim=1).mean().neg()
 
-            total_loss += mutual_loss / float(num_mutual_losses)
+                        mutual_loss += m_loss
+                        loss_summary['mutual_{}/{}'.format(trg_id, model_names[model_id])] = m_loss.item()
+                        num_mutual_losses += 1
 
-        total_loss.backward()
+                total_loss += mutual_loss / float(num_mutual_losses)
 
-        for model_name in model_names:
-            for trg_id in range(self.num_targets):
-                if self.enable_metric_losses:
-                    ml_loss_module = self.ml_losses[trg_id][model_name]
-                    ml_loss_module.end_iteration(do_backward=False)
+            total_loss.backward(retain_graph=self.enable_metric_losses)
 
-            self.optims[model_name].step()
+            for model_name in model_names:
+                optim_name = self.optims[model_name].__class__.__name__
+                for trg_id in range(self.num_targets):
+                    if self.enable_metric_losses:
+                        ml_loss_module = self.ml_losses[trg_id][model_name]
+                        ml_loss_module.end_iteration(do_backward=False)
+                if optim_name == "SAM" and step == 1:
```

Comment:
I'd rather use `isinstance` here, but it's a hack anyway. Why not to check `self.enable_sam`?